operator/pkg/controller/karmada/controller.go

```diff
@@ -61,38 +61,28 @@ func (ctrl *Controller) Reconcile(ctx context.Context, req controllerruntime.Req
 				return controllerruntime.Result{}, err
 			}
 		}
-	} else {
-		// The object is being deleted
-		if controllerutil.ContainsFinalizer(karmada, ControllerFinalizerName) {
-			// our finalizer is present, so lets handle any external dependency
-			if err := ctrl.deleteUnableGCResources(karmada); err != nil {
-				// if fail to delete the external dependency here, return with error
-				// so that it can be retried
-				return controllerruntime.Result{}, err
-			}
-
-			// remove our finalizer from the list and update it.
-			controllerutil.RemoveFinalizer(karmada, ControllerFinalizerName)
-			if err := ctrl.Update(ctx, karmada); err != nil {
-				return controllerruntime.Result{}, err
-			}
-			// Stop reconciliation as the item is being deleted
-			return controllerruntime.Result{}, nil
-		}
 	}
 
 	klog.V(2).InfoS("Reconciling karmada", "name", req.Name)
-
 	planner, err := NewPlannerFor(karmada, ctrl.Client, ctrl.Config)
 	if err != nil {
 		return controllerruntime.Result{}, err
 	}
-	return planner.Execute()
-}
+	if err := planner.Execute(); err != nil {
+		return controllerruntime.Result{}, err
+	}
+
+	// The object is being deleted
+	if !karmada.DeletionTimestamp.IsZero() && controllerutil.ContainsFinalizer(karmada, ControllerFinalizerName) {
+		// remove our finalizer from the list and update it.
+		controllerutil.RemoveFinalizer(karmada, ControllerFinalizerName)
```

Comment:
Suggest to encapsulate it as a function.

like

https://github.com/karmada-io/karmada/blob/a4d3eced85ede213c9c6cf6813ea564dde1b6583/pkg/controllers/execution/execution_controller.go#L171-L183

Reply:
ok